### ISSUE TITLE
Ambilight not working on RPi2 + Arduino with WS2812b Strip

### ISSUE THREAD

**Issue by Hibuja** [tvdzwan/hyperion#380]
Hi guys,
I've been trying to get Ambilight to work for about a week now (I spent at least 40h trying to get it to work) but whatever I do it just won't work. At first I tried it directly with the Pi but soon found out that it wouldn't work and got out my Arduino.
Here's a list of what I've been trying without any success:
- I tried 3 different Arduino skeches on 2 different Arduinos (Mega 2560 and Nano) with Neopixel as well as FastLED library (Both libraries work with examples only on the arduino)
- I tried a RPi 1 and RPi 2 (both only with OpenELEC & Kodi, not Raspbian)
- I tried Hyperion as well as Boblight
- I fed the Arduino with data by hand to ensure correctly working code with correct data (and it worked)
- I captured the data sent by boblight (the data was correct)
- I can control the strip with the arduino so I assume there is no problem with the wiring
- I tried sending the data from the Pi to the Arduino via USB and UART. Both seemed to work because the Arudinos received the data (RX led flashing) but the LED Strip stayed dark
- I tried changing the baudrate from 100000 to 500000 but no effect

I got it working one time with boblight but after a reboot it stopped working again and ~10 Pi reboots, ~20 boblight restarts and like 50 arduino resets couldn't bring it back to a working state.
Here you can find my config file for 120 LEDs http://pastebin.com/wYL0M6Y4
I had to disable the XBMC grabber because there was an xbmc connection error.

Here is the console output when I start hyperion via SSH: http://pastebin.com/AsxHwf2C

The only option I see is to change over to Raspbian and Kodi but I tried to avoid this as long as possible because everything except Ambilight is setup and working fine. If anyone has any idea what the problem could be please let me know or if you need any more info about the wiring or anything just tell me and I will post it.

Any help is really appreciated! :)

**Comment by Hibuja** [comment]
Ok I just got it working pretty well. There were actually a few problems which I will list here so if anyone has the same problems he or she could check these issues.

1. I had the wrong prefix. I used the prefix the boblight config tool generated (41 64 61 00 78 2D). But this prefix differs from the prefix used by hyperion (41 64 61 00 77 22) as hyperion uses NUM_LEDS-1 as byte 4 and 5 and the boblight config tool uses the actual number of leds so there was a difference of one between those bytes and then the checksum was obviously wrong too.

2. I can't use the USB Port for communication. I don't know why but the data coming from USB were total rubbish and incomprehensible. Sp I changed over to UART on Pin 8 of RPi 2.

3. I had to lower the baudrate. The first working baudrate was 115200. All tested baudrates (150000, 230400 and 460800) above this one result in some kind of communication error and the arduino can't read the data.

The only thing left is that the XBMC grabber still says that there is a connection error. So any suggestions what that might be due to?

**Comment by AlexBGoode** [comment]
I also wasted a lot of time trying to get it running... I managed it finally. And yes I would confirm that it is really strange to calculate the Ada-prefix using not the number of LEDs but the index of the last led, which is the number of LEDs minus one!!! I changed everything back and fourth until downloaded a pair of hyperion config / arduino sketch (FastLED) reported as working well and it worked for me as well but only for 72 LEDs... I discovered that all successful projects use an inconsistent configuration when the Ada-prefix is calculated for one less count of LEDs... so much time wasted figuring it out.
It is a strange behaviour I have to say.

So in my case I have 220 LEDs and the arduino sketch uses "Ada 0x00, 0xDB, 0x8E" prefix which means 219!

I think it would be very useful if hyperiond reports the number of configured LEDs and the Ada-prefix used on startup for debugging purposes... as for now I have only guess which exactly prefix it uses
